### PR TITLE
Unsupported keys per provider

### PR DESCRIPTION
**Pull request by rtnpro**
Separate unsupported keys based on provider.

### CODE REVIEW

**Comment by rtnpro** [comment]
#206 is blocked on this.

**Comment by sebgoa** [comment]
Can we have unit tests for this, please ?

**Comment by kadel** [comment]
@sebgoa for sure. It is not going to be merged without it.

File: pkg/kobject/kobject.go [hunk 1]
```diff
@@ -16,93 +16,13 @@ limitations under the License.
...
-	"Net":           "net",
-	"Args":          "args",
-}
+import "k8s.io/kubernetes/pkg/api"
 
 // KomposeObject holds the generic struct of Kompose transformation
 type KomposeObject struct {
 	ServiceConfigs map[string]ServiceConfig
+	// name of the loader that was created KomposeObject
+	LoadedFrom string
```
Comment (surajssd) [inline pkg/kobject/kobject.go]: why is this required?

File: pkg/kobject/kobject.go [hunk 2]
```diff
@@ -122,28 +42,30 @@ type ConvertOptions struct {
...
+	CPUQuota      int64             `compose:"cpu_quota",bundle:""`
+	CapAdd        []string          `compose:"cap_add",bundle:""`
+	CapDrop       []string          `compose:"cap_drop",bundle:""`
+	Expose        []string          `compose:"expose",bundle:""`
+	Privileged    bool              `compose:"privileged",bundle:""`
+	Restart       string            `compose:"restart",bundle:""`
+	User          string            `compose:"user",bundle:"User"`
+	VolumesFrom   []string          `compose:"volumes_from",bundle:""`
+	ServiceType   string            `compose:"kompose.service.type",bundle:""`
+	Build         string            `compose:"build",bundle:""`
```
Comment (surajssd) [inline pkg/kobject/kobject.go]: why are tags needed for ServiceConfig ? Is this for `structs`?

File: pkg/kobject/kobject.go (hunk 1 above)
Comment (kadel) [inline pkg/kobject/kobject.go]: Transformer need to know origin format in order to tell user what tag is not supported in origin format. They can have different names.
for example environment  variables. are called  `environment` in compose,`Env` in bundle.

File: pkg/loader/compose/compose.go [hunk 3]
```diff
@@ -29,12 +30,107 @@ import (
...
+				// IsZero returns false for empty array/slice ([])
+				// this check if field is Slice, and then it checks its size
+				if field := val.FieldByName(f.Name()); field.Kind() == reflect.Slice {
+					if field.Len() == 0 {
+						// array is empty it doesn't metter if it is in unsupportedKey or not
+						continue
+					}
+				}
+
+				if counter, ok := unsupportedKey[f.Name()]; ok {
```
Comment (surajssd) [inline pkg/loader/compose/compose.go]: Because this particular check decides if that particular key is supported or not?

File: pkg/loader/bundle/bundle.go [hunk 4]
```diff
@@ -57,6 +59,53 @@ type Port struct {
...
+				jsonTagName := strings.Split(f.Tag("json"), ",")[0]
+				if jsonTagName == "" {
+					jsonTagName = f.Name()
+				}
+
+				// IsZero returns false for empty array/slice ([])
+				// this check if field is Slice, and then it checks its size
+				if field := val.FieldByName(f.Name()); field.Kind() == reflect.Slice {
+					if field.Len() == 0 {
+						// array is empty it doesn't metter if it is in unsupportedKey or not
```
Comment (surajssd) [inline pkg/loader/bundle/bundle.go]: typo: `s/metter/matter/`

File: pkg/transformer/kubernetes/kubernetes.go [hunk 5]
```diff
@@ -51,6 +53,52 @@ type Kubernetes struct {
...
+		val := reflect.ValueOf(serviceConfig)
+		s := structs.New(serviceConfig)
+
+		for _, f := range s.Fields() {
+			if f.IsExported() && !f.IsZero() {
+				// IsZero returns false for empty array/slice ([])
+				// this check if field is Slice, and then it checks its size
+				if field := val.FieldByName(f.Name()); field.Kind() == reflect.Slice {
+					if field.Len() == 0 {
+						// array is empty it doesn't metter if it is in unsupportedKey or not
```
Comment (surajssd) [inline pkg/transformer/kubernetes/kubernetes.go]: typo: `s/metter/matter/`

File: pkg/loader/compose/compose.go [hunk 6]
```diff
@@ -29,12 +30,107 @@ import (
...
+		val := reflect.ValueOf(serviceConfig).Elem()
+		s := structs.New(serviceConfig)
+
+		for _, f := range s.Fields() {
+			if f.IsExported() && !f.IsZero() {
+				// IsZero returns false for empty array/slice ([])
+				// this check if field is Slice, and then it checks its size
+				if field := val.FieldByName(f.Name()); field.Kind() == reflect.Slice {
+					if field.Len() == 0 {
+						// array is empty it doesn't metter if it is in unsupportedKey or not
```
Comment (surajssd) [inline pkg/loader/compose/compose.go]: typo: `s/metter/matter/`

File: pkg/loader/compose/compose.go (hunk 3 above)
Comment (kadel) [inline pkg/loader/compose/compose.go]: Technically It could be moved there, I kept it in the same order as it was before.

I wouldn't say a lot of processing as happening before. It is all just simple checks. Plus it is not like there will ever be hundreds of keys.
But I updated it anyway ;-)  Plus I changed int counter in unsuportedKeys to bool. 
Because all we need is to know if there already was that key or not ;-)

File: pkg/transformer/kubernetes/kubernetes.go (hunk 5 above)
Comment (kadel) [inline pkg/transformer/kubernetes/kubernetes.go]: fixed

File: pkg/loader/compose/compose.go (hunk 6 above)
Comment (kadel) [inline pkg/loader/compose/compose.go]: fixed

**Comment by surajssd** [comment]
@ngtuna your say here will be appreciated!

**Comment by rtnpro** [comment]
@surajssd rebased and ready for merging.

**Comment by surajssd** [comment]
@rtnpro @kadel  Thanks :+1:
